plugins/statistics.koplugin/main.lua

```diff
@@ -328,34 +445,67 @@ function ReaderStatistics:createDB(conn)
                 total_read_time  integer,
                 total_read_pages integer
             );
-        CREATE TABLE IF NOT EXISTS page_stat
-            (
-                id_book    integer,
-                page       integer NOT NULL,
-                start_time integer NOT NULL,
-                period     integer NOT NULL,
-                UNIQUE (page, start_time),
-                FOREIGN KEY(id_book) REFERENCES book(id)
-             );
-        CREATE TABLE IF NOT EXISTS info
-             (
-                 version integer
-             );
-        CREATE INDEX IF NOT EXISTS page_stat_id_book ON page_stat(id_book);
+    ]]
+    conn:exec(sql_stmt)
+
+    -- page_stat_data
+    conn:exec(STATISTICS_DB_PAGE_STAT_DATA_SCHEMA)
+
+    sql_stmt = [[
+        -- Indexes
+        CREATE INDEX IF NOT EXISTS page_stat_data_id_book ON page_stat_data(id_book);
         CREATE INDEX IF NOT EXISTS book_title_authors_md5 ON book(title, authors, md5);
     ]]
     conn:exec(sql_stmt)
-    --DB structure version - now is version 1
-    local stmt = conn:prepare("INSERT INTO info values (?)")
-    stmt:reset():bind("1"):step()
-    stmt:close()
+
+    -- page_stat view
+    conn:exec(STATISTICS_DB_PAGE_STAT_VIEW_SCHEMA)
+
+    -- DB schema version
+    conn:exec(string.format("PRAGMA user_version=%d;", DB_SCHEMA_VERSION))
+end
+
+function ReaderStatistics:upgradeDB(conn)
+    local sql_stmt = [[
+        -- Start by updating the layout of the old page_stat table
+        ALTER TABLE page_stat RENAME COLUMN period TO duration;
+        -- We're now using the user_version PRAGMA to keep track of schema version
+        DROP TABLE info;
+    ]]
+    conn:exec(sql_stmt)
+
+    -- Migrate page_stat content to page_stat_data, which we'll have to create first ;).
+    conn:exec(STATISTICS_DB_PAGE_STAT_DATA_SCHEMA)
+
+    sql_stmt = [[
+        -- Migrate page_stat content to page_stat_data, and populate total_pages from book's pages while we're at it.
+        -- NOTE: While doing a per-book migration could ensure a potentially more accurate page count,
+        --       we need to populate total_pages *now*, or queries against unopened books would return completely bogus values...
+        --       We'll just have to hope the current value of the column pages in the book table is not too horribly out of date,
+        --       and not too horribly out of phase with the actual page count at the time the data was originally collected...
+        INSERT INTO page_stat_data
+            SELECT id_book, page, start_time, duration, pages as total_pages FROM page_stat
+            LEFT JOIN book on book.id = id_book;
+
+        -- Update the index, too
+        DROP INDEX IF EXISTS page_stat_id_book;
```

Comment:
I wanted to drop the index *after* the table, because I suppose it's still useful during the migration?

(Remember, I suck at SQL ;o)).

Reply:
Yes, it would still be after :) "Move this DROP INDEX _before_ the DROP TABLE _below_":

```
        CREATE INDEX IF NOT EXISTS page_stat_data_id_book ON page_stat_data(id_book);

        -- Drop old page_stat table
        DROP INDEX IF EXISTS page_stat_id_book;
        DROP TABLE IF EXISTS page_stat;
```

Reply:
Oh. :D.